dough-inventories/src/main/java/io/github/bakedlibs/dough/inventory/factory/CustomMenu.java

```diff
@@ -0,0 +1,141 @@
+
+    @ParametersAreNonnullByDefault
+    protected CustomMenu(MenuFactory factory, MenuLayout layout) {
+        Validate.notNull(factory, "The factory cannot be null.");
+        Validate.notNull(layout, "The layout cannot be null.");
+
+        this.factory = factory;
+        this.layout = layout;
+        this.title = layout.getTitle();
+    }
+
+    final void setInventory(@Nonnull Inventory inventory) {
+        Validate.notNull(inventory, "The Inventory must not be null.");
+        Validate.isTrue(inventory.getSize() == layout.getSize(), "The inventory has a different size.");
+        Validate.isTrue(inventory.getHolder() == this, "The Inventory does not seem to belong here. Holder: " + inventory.getHolder());
+
+        this.inventory = inventory;
+    }
+
+    private void validate() {
+        if (inventory == null) {
+            throw new UnsupportedOperationException("No inventory found! Menus must be created using MenuFactory#createMenu(...)");
+        }
+    }
+
+    @Override
+    public final @Nonnull MenuFactory getFactory() {
+        validate();
+
+        return factory;
+    }
+
+    @Override
+    public final @Nonnull Inventory getInventory() {
+        validate();
+
+        return inventory;
+    }
+
+    @Override
+    public final @Nonnull MenuLayout getLayout() {
+        validate();
+
+        return layout;
+    }
+
+    @Override
+    public String getTitle() {
+        validate();
+
+        return title;
+    }
+
+    @Override
+    public void setAll(SlotGroup group, ItemStack item) {
+        validate();
+
+        if (group.size() == 1) {
+            // Little optimization, we don't need to create an Iterator for this
+            setItem(group.getSlots()[0], item);
+        } else {
+            for (int slot : group) {
+                setItem(slot, item);
+            }
+        }
+    }
+
+    @Override
+    @ParametersAreNonnullByDefault
+    public @Nullable ItemStack addItem(SlotGroup group, ItemStack item) {
+        Validate.notNull(group, "The Slot group cannot be null!");
+        Validate.notNull(item, "The Item cannot be null!");
+        validate();
+
+        for (int slot : group) {
+            ItemStack itemInSlot = getItem(slot);
+
+            if (itemInSlot == null || itemInSlot.getType().isAir()) {
+                setItem(slot, item);
+                return null;
+            } else {
+                int currentAmount = itemInSlot.getAmount();
+                int maxStackSize = itemInSlot.getType().getMaxStackSize();
+
+                if (currentAmount < maxStackSize && itemInSlot.isSimilar(item)) {
+                    int amount = currentAmount + item.getAmount();
+
+                    if (amount > maxStackSize) {
+                        item.setAmount(amount - maxStackSize);
+                        itemInSlot.setAmount(maxStackSize);
+                    } else {
+                        itemInSlot.setAmount(Math.min(amount, maxStackSize));
```

Comment:
You can forget about the Math#min here for a slight gain in speed and just set to "amount": you have already checked that "amount" is at most maxStackSize so worst case scenario is the two are equal and Math.min will set exactly "amount" anyway, otherwise "amount" is the lower so it would again be the chosen one